imap/imap.go

```diff
@@ -0,0 +1,630 @@
+	return generateMail(info, "ALL", nil, markAsRead, delete)
+}
+
+// GetUnread will find all unread emails in the folder and return them as a list.
+func GetUnread(info MailboxInfo, markAsRead, delete bool) ([]Email, error) {
+	// call chan, put 'em in a list, return
+	var emails []Email
+
+	responses, err := GenerateUnread(info, markAsRead, delete)
+	if err != nil {
+		return emails, err
+	}
+
+	for resp := range responses {
+		if resp.Err != nil {
+			return emails, resp.Err
+		}
+		emails = append(emails, resp.Email)
+	}
+
+	return emails, nil
+}
+
+// GenerateUnread will find all unread emails in the folder and pass them along to the responses channel.
+func GenerateUnread(info MailboxInfo, markAsRead, delete bool) (chan Response, error) {
+	return generateMail(info, "UNSEEN", nil, markAsRead, delete)
+}
+
+// GetSince will pull all emails that have an internal date after the given time.
+func GetSince(info MailboxInfo, since time.Time, markAsRead, delete bool) ([]Email, error) {
+	var emails []Email
+	responses, err := GenerateSince(info, since, markAsRead, delete)
+	if err != nil {
+		return emails, err
+	}
+
+	for resp := range responses {
+		if resp.Err != nil {
+			return emails, resp.Err
+		}
+		emails = append(emails, resp.Email)
+	}
+
+	return emails, nil
+}
+
+// GenerateSince will find all emails that have an internal date after the given time and pass them along to the
+// responses channel.
+func GenerateSince(info MailboxInfo, since time.Time, markAsRead, delete bool) (chan Response, error) {
+	return generateMail(info, "", &since, markAsRead, delete)
+}
+
+// MarkAsUnread will set the UNSEEN flag on a supplied slice of UIDs
+func MarkAsUnread(info MailboxInfo, uids []uint32) error {
+
+	client, err := newIMAPClient(info)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		client.Close(true)
+		client.Logout(30 * time.Second)
+	}()
+	for _, u := range uids {
+		err := alterEmail(client, u, "\\SEEN", false)
+		if err != nil {
+			return err //return on first failure
+		}
+	}
+	return nil
+
+}
+
+// DeleteEmails will delete emails from the supplied slice of UIDs
+func DeleteEmails(info MailboxInfo, uids []uint32) error {
+
+	client, err := newIMAPClient(info)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		client.Close(true)
+		client.Logout(30 * time.Second)
+	}()
+	for _, u := range uids {
+		err := deleteEmail(client, u)
+		if err != nil {
+			return err //return on first failure
+		}
+	}
+	return nil
+
+}
+
+// ValidateIMAP validates supplied IMAP model by connecting to the server
+func ValidateIMAP(s *models.IMAP) error {
+
+	err := s.Validate()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	s.Host = s.Host + ":" + strconv.Itoa(int(s.Port)) // Append port
+	mailSettings := MailboxInfo{
+		Host:   s.Host,
+		TLS:    s.TLS,
+		User:   s.Username,
+		Pwd:    s.Password,
+		Folder: s.Folder}
+
+	client, err := newIMAPClient(mailSettings)
+	defer func() {
+		client.Close(true)
```

Comment:
By calling this every time the function runs, you run the risk of causing crashes if `err != nil`. In my case, during testing I gave invalid IMAP settings which threw an error, setting `client` to `nil`. This caused a crash since we're trying to call `nil.Close()`.

Instead, in this case we should check the error first, returning it if it's not nil and then either call the defer function or, probably easier here, just run those cleanup functions directly since there isn't any code to run after error checking.